tools/src/launcher/Main.hs

```diff
@@ -479,46 +505,88 @@ clientScenario ndbp logConf node wallet updater nodeTimeout report walletLog = d
-                        removeInstaller = removeFile updateArchivePath
-                    -- Even if we fail to affirm that update was
-                    -- installed, we still want to remove installer to
-                    -- avoid infinite loop. If we don't remove it, we
-                    -- will launch it again and again.
-                    affirmInstalled `finally` removeInstaller
-            ExitFailure code ->
-                logWarning $ sformat ("The updater has failed (exit code "%int%")") code
+        hashValid <- validInstallerHash installerPath
+        if hashValid
+            then do
+                exitCode <- runUpdaterWithOS ud
+                archiveAndRemove exitCode installerPath
+            else do
+                logWarning "Installer failed Hash check and will be removed"
+                liftIO $ removeFile installerPath
+
+  where
+    validInstallerHash :: FilePath -> M Bool
+    validInstallerHash instPath = do
+
+        validHash <- liftIO $ bracketNodeDBs ndbp $ \lmcNodeDBs ->
+            usingReaderT LauncherModeContext{..} $ do
+                getLastInstallerHash
+
+        installer <- liftIO $ BS.L.readFile instPath
+        let currentHash = installerHash installer
+        return $ currentHash == validHash 
+
+    getInstaller :: UpdaterData -> FilePath
+    getInstaller (WindowsUpdaterData WinUD {..}) = wudExePath
+    getInstaller (OSXUpdaterData OSXUD {..}) = oudPkgPath
+    -- Should not happen due to whenM in runUpdater
+    getInstaller _ = error "Unsupported OS UpdateConfig has no Installer"
+
+    archiveAndRemove :: ExitCode -> FilePath -> M ()
+    archiveAndRemove exitCode installerPath = case exitCode of
+        ExitSuccess -> do
+            logInfo "The updater has exited successfully"
+            -- this will throw an exception if the file doesn't exist but
+            -- hopefully if the updater has succeeded it *does* exist
+            liftIO $ do 
+                let affirmInstalled = do
+                        updateArchive <- BS.L.readFile installerPath
+                        bracketNodeDBs ndbp $ \lmcNodeDBs ->
+                            usingReaderT LauncherModeContext{..} $
+                            affirmUpdateInstalled (installerHash updateArchive)
+                    removeInstaller = removeFile installerPath
+                -- Even if we fail to affirm that update was
+                -- installed, we still want to remove installer to
+                -- avoid infinite loop. If we don't remove it, we
+                -- will launch it again and again.
+                affirmInstalled `finally` removeInstaller
+        ExitFailure code ->
+            logWarning $ sformat ("The updater has failed (exit code "%int%")") code
+
+    logIfM predicate message = ifM predicate (return True) ((logWarning message) >> return False)
+
+    logIf predicate message = if predicate 
+        then return True
+        else (logWarning message) >> return False
```

Comment:
Btw, `>> return` is `$>` (i. e. this line can be written as `logWarning message $> False`), but it's up to you to decide which version is better.